### PR TITLE
Model() yield op Input internal fixes, plus add K.is_placeholder()

### PR DESCRIPTION
**Pull request by ahundt**
These changes include `K.is_placeholder()`, plus the internal changes of #6928, excluding the `labels` parameter or TFRecord support. This includes better checks to prevent data from being fed to `feed_dict` or `fetches` incorrectly during model.run(), improved user input sanity checks, and simplified backend function initialization.

Please also consider my alternative external API proposals in https://github.com/fchollet/keras/pull/6928#issuecomment-309595218, but that should be kept separate from these internal improvements.

### CODE REVIEW

**Comment by ahundt** [comment]
@Dref360 You'll be familiar with all of these changes, so I'd appreciate if you could give it another look.

**Comment by ahundt** [comment]
Failure of 3be24ba is due to https://github.com/fchollet/keras/pull/7030

**Comment by ahundt** [comment]
tests passed except #7030

**Comment by ahundt** [comment]
only failure was flaky test https://github.com/fchollet/keras/pull/7033, closing and reopening

**Comment by ahundt** [comment]
only failure was flaky test #7033, closing and reopening

File: examples/mnist_tfrecord.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+y_train_in_out = Input(tensor=y_train_batch, batch_shape=y_batch_shape, name='y_labels')
+cce = categorical_crossentropy(y_train_batch, x_train_out)
+train_model = Model(inputs=[x_train_input], outputs=[x_train_out])
+train_model.add_loss(cce)
+
+train_model.compile(optimizer='rmsprop',
+                    loss=None,
+                    metrics=['accuracy'])
+
+tensorboard = TensorBoard(write_graph=True)
```
Comment (Dref360) [inline examples/mnist_tfrecord.py]: useless?

File: examples/mnist_tfrecord.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+x_batch_shape = x_train_batch.get_shape().as_list()
+y_batch_shape = y_train_batch.get_shape().as_list()
+
+x_train_input = Input(tensor=x_train_batch, batch_shape=x_batch_shape)
+x_train_out = cnn_layers(x_train_input)
+y_train_in_out = Input(tensor=y_train_batch, batch_shape=y_batch_shape, name='y_labels')
+cce = categorical_crossentropy(y_train_batch, x_train_out)
+train_model = Model(inputs=[x_train_input], outputs=[x_train_out])
+train_model.add_loss(cce)
```
Comment (Dref360) [inline examples/mnist_tfrecord.py]: I don't think a lot of people knew about this trick. Really nice!

File: examples/mnist_tfrecord.py (hunk 1 above)
Comment (ahundt) [inline examples/mnist_tfrecord.py]: I had to disable it because without the `expected_targets` param, (I'd like to rename it to that from `labels`) much of Keras' functionality is broken in this use case. Passing `callbacks=[tensorboard]` crashes. It runs perfectly in #6928, with limited functionality with this PR, and on master, not at all...

File: examples/mnist_tfrecord.py (hunk 2 above)
Comment (ahundt) [inline examples/mnist_tfrecord.py]: credit: @fchollet 

However, it is extremely limiting, a good portion of the functionality enabled by model.compile() is skipped...

**Comment by ahundt** [comment]
I've split this up into many smaller PRs, the capstone change is https://github.com/fchollet/keras/pull/7072 which should link to all the smaller ones.

File: examples/mnist_tfrecord.py (hunk 2 above)
Comment (ahundt) [inline examples/mnist_tfrecord.py]: Additionally I've discovered this trick has serious performance problems, see https://github.com/fchollet/keras/pull/7075#issue-237440526. It is **20x** slower than `mnist_tfrecord.py` in #6928

**Comment by ahundt** [comment]
This PR has been superseded by #6928 because of https://github.com/fchollet/keras/issues/7102#issuecomment-310809267. For that reason I'm closing this PR.
